src/stream/src/executor/gap_fill.rs

```diff
@@ -0,0 +1,1856 @@
+                return Ok(filled_rows);
+            }
+        };
+
+        if prev_time >= curr_time {
+            return Ok(filled_rows);
+        }
+
+        // Generate template row with fill strategies
+        let mut fill_values: Vec<Datum> = Vec::with_capacity(prev_row.len());
+        for i in 0..prev_row.len() {
+            if i == time_column_index {
+                fill_values.push(None); // Will be set later
+            } else if let Some((_, strategy)) = fill_columns.iter().find(|(col, _)| *col == i) {
+                match strategy {
+                    FillStrategy::Locf | FillStrategy::Interpolate => {
+                        fill_values.push(prev_row.datum_at(i).to_owned_datum())
+                    }
+                    FillStrategy::Null => fill_values.push(None),
+                }
+            } else {
+                fill_values.push(prev_row.datum_at(i).to_owned_datum());
+            }
+        }
+
+        // Generate filled timestamps
+        let mut fill_time = match prev_time.checked_add(*interval) {
+            Some(t) => t,
+            None => return Ok(filled_rows),
+        };
+
+        let mut data = Vec::new();
+        while fill_time < curr_time {
+            let mut new_row_data = fill_values.clone();
+            let fill_time_scalar = match prev_time_scalar {
+                ScalarRefImpl::Timestamp(_) => ScalarImpl::Timestamp(fill_time),
+                ScalarRefImpl::Timestamptz(_) => {
+                    let micros = fill_time.0.and_utc().timestamp_micros();
+                    ScalarImpl::Timestamptz(risingwave_common::types::Timestamptz::from_micros(
+                        micros,
+                    ))
+                }
+                _ => unreachable!("Time column should be Timestamp or Timestamptz"),
+            };
+            new_row_data[time_column_index] = Some(fill_time_scalar);
+            data.push(new_row_data);
+
+            fill_time = match fill_time.checked_add(*interval) {
+                Some(t) => t,
+                None => break,
+            };
+        }
+
+        // --- NEW LOGIC FOR INTERPOLATION ---
+        for (col_idx, strategy) in fill_columns {
+            if matches!(strategy, FillStrategy::Interpolate) {
+                let steps = data.len();
+                let step = Self::calculate_step(
+                    prev_row.datum_at(*col_idx).to_owned_datum(),
+                    curr_row.datum_at(*col_idx).to_owned_datum(),
+                    steps + 1,
+                );
+                if let Some(step) = step {
+                    let mut cumulative_value = prev_row.datum_at(*col_idx).to_owned_datum();
+                    for row in &mut data {
+                        Self::apply_step(&mut cumulative_value, &step);
+                        row[*col_idx] = cumulative_value.clone();
+                    }
+                }
+            }
+        }
```

Comment:
Seems like these two logics can be unified. As we generate a new row, we apply a function to generate each datum in the row. This function takes in: previous datum, strategy (timestamp increment, locf, null, interpolate), time interval, upper bound datum.